cmd/kubeadm/app/util/kustomize/kustomize.go

```diff
@@ -0,0 +1,181 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package kustomize contains helpers for working with embedded kustomize commands
+package kustomize
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"runtime"
+	"sync"
+
+	"k8s.io/cli-runtime/pkg/kustomize"
+	"sigs.k8s.io/kustomize/pkg/fs"
+	"sigs.k8s.io/kustomize/pkg/loader"
+)
+
+// Manager define a manager that allow access to kustomize capabilities
+type Manager struct {
+	kustomizeDir string
+	us           UnstructuredSlice
+}
+
+var (
+	lock      = &sync.Mutex{}
+	instances = map[string]*Manager{}
+)
+
+// GetManager return the KustomizeManager singleton instance
+func GetManager(kustomizeDir string) (*Manager, error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	// if the instance does not exists, create it
+	if _, ok := instances[kustomizeDir]; !ok {
+		km := &Manager{
+			kustomizeDir: kustomizeDir,
+		}
+
+		// loads the UnstructuredSlice with all the patches into the Manager
+		// NB. this is done at singleton instance level because kubeadm has a unique pool
+		// of patches that are applied to different content, at different time
+		if err := km.getUnstructuredSlice(); err != nil {
+			return nil, err
+		}
+
+		instances[kustomizeDir] = km
+	}
+
+	return instances[kustomizeDir], nil
+}
+
+// getUnstructuredSlice returns a UnstructuredSlice with all the patches.
+func (km *Manager) getUnstructuredSlice() error {
+	// kubeadm does not require a kustomization.yaml file listing all the resources/patches, so it is necessary
+	// to rebuild the list of patches manually
+	// TODO: make this git friendly - currently this works only for patches in local folders -
+	files, err := ioutil.ReadDir(km.kustomizeDir)
+	if err != nil {
+		return err
+	}
+
+	var paths = []string{}
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		paths = append(paths, file.Name())
+	}
+
+	// Create a loader that mimics the behavior of kubectl kustomize, inclusing support for reading from
+	// a local git repository like git@github.com:someOrg/someRepo.git or https://github.com/someOrg/someRepo?ref=someHash
+	fSys := fs.MakeRealFS()
+	ldr, err := loader.NewLoader(km.kustomizeDir, fSys)
+	if err != nil {
+		return err
+	}
+	defer ldr.Cleanup()
+
+	// read all the kustomizations and build the UnstructuredSlice
+	us, err := NewUnstructuredSliceFromFiles(ldr, paths)
+	if err != nil {
+		return err
+	}
+
+	km.us = us
+	return nil
+}
+
+// Kustomize apply a set of patches to a resource.
+// Portions of the kustomize logic in this function are taken from the kubernetes-sigs/kind project
+func (km *Manager) Kustomize(res []byte) ([]byte, error) {
+	// create a loader that mimics the behavior of kubectl kustomize
```

Comment:
what ensures this mimicked behavior doesn't drift from actual kustomize behavior?

Reply:
cc @monopole for insights on more consistent ways to make use of the kustomize capabilities

Reply:
Just a note waiting for @monopole review
Being kubeadm still in the main k/k repo, we are not allowed to import "k8s.io/cli-runtime/pkg/kustomize/k8sdeps", when kubeadm will move out from k/k (ETA 1.18), there will be room to further reduce risk of drift by using types defined in this package.

Reply:
Sorry about the delay, i was out for a bit.

There’s no formal notion of kustomize as a versioned library or service as yet, though it’s being worked on.  So the kustomize API is merely 1) the main program command line arguments and 2) the kustomization file and patches it references.

Ergo the recommended way to integrate kustomize into program foo is to have foo write a kustomization file and associated resources / patches to a file system, execute kustomize in a sub-process with the desired working directory, then capture stdout for whatever purpose (discarding it on error).  This is how kubebuilder uses it.  One can upgrade kustomize (the binary) without recompiling foo (modulo major version numbers).

This PR looks to be integrating kustomize into kubeadm in a fashion similar to how kustomize is integrated into kubectl, via the snapshot at k8s.io/cli-runtime/pkg/kustomize.  Doing this means kubeadm users don't have to install kustomize, but you’re tying yourself to 1) an old kustomize snapshot, and 2) _a code arrangement intended only for kubectl and not supported for others._

I.e. the only build we'll try not to break as we change cli-runtime/pkg/kustomize is the kubectl build.  So at some point the kubeadm build will break.  I don't think the fix will be hard (some new imports, new call styles, etc. without much change to overall concept), but it will presumably be annoying.

Reply:
@monopole thanks for the valuable feedback!
My take away is that the current implementation has some trade-off that can be acceptable for the alpha release of this feature, but in the long term we should converge on one of the kubebuilder or the kubectl model, or be a more active part of the ongoing work around "kustomize as a versioned library". I'm open an issue in the k/kustomize repo to properly track this.

Unless there are no different opinions, I'm +1 to get this merged as it is because the above points are related only to implementation details, without any impact on the user experience

Reply:
as per comment above https://github.com/kubernetes-sigs/kustomize/issues/1441